mirascope/gemini/tools.py

```diff
@@ -68,20 +73,42 @@ def tool_schema(cls) -> Tool:
         Returns:
             The constructed `Tool` schema.
         """
-        tool_schema = super().tool_schema()
-        if "parameters" in tool_schema:
-            if "$defs" in tool_schema["parameters"]:
-                raise ValueError(
-                    "Unfortunately Google's Gemini API cannot handle nested structures "
-                    "with $defs."
+        super().tool_schema()
+        model_schema: dict[str, Any] = cls.model_json_schema()
+        pprint.pprint(model_schema)
+
+        # Replace all references with their values
+        without_refs: dict[str, Any] = jsonref.replace_refs(model_schema)  # type: ignore
+        pprint.pprint(without_refs)
+
+        # Remove all Defs
+        without_refs.pop("$defs")
+        pprint.pprint(without_refs)
+
+        # Get the name and description, and remove them from the schema
+        name: str = without_refs.pop("title")  # type: ignore
+        description: str = (  # type: ignore
+            without_refs.pop("description", None) or DEFAULT_TOOL_DOCSTRING
+        )
+        parameters: dict[str, Any] = without_refs
+
+        # Remove all instances of title key in each param definition
+        # This is careful not to delete keys that represent a field with the name title
+        remove_invalid_title_keys_from_parameters(parameters)
```

Comment:
what if all of the conversion of `parameters` was pushed inside this utility function so the main function remains clean and the utility function can be separately unit tested?